File: apps/web/src/types/plan.ts
import type {
  PlanStepStatus,
  PlanStepTool,
  RiskLevel,
} from './enums'
import type { RepoCandidate } from './repoDiscovery'

export const DEFAULT_RUBRIC_WEIGHTS: Record<string, number> = {
  reproducibility: 0.3,
  documentation: 0.2,
  research_value: 0.2,
  engineering_fit: 0.2,
  extensibility: 0.1,
}

export interface ClarifyingQuestion {
  id: string
  question: string
  options: string[]
  answer: string | null
  required: boolean
}

export interface PlanStep {
  id: string
  node_id: string  // F-002: matches WorkflowNode in execute graph
  title: string
  description: string
  tool: PlanStepTool
  risk_level: RiskLevel
  requires_approval: boolean
  expected_outputs: string[]
  depends_on: string[]
  user_editable: boolean
  status: PlanStepStatus
}

export interface ReportRubric {
  weights: Record<string, number>
}

export interface Plan {
  id: string
  task_id: string
  task_summary: string
  confirmed_context: string[]
  unknowns: string[]
  clarifying_questions: ClarifyingQuestion[]
  proposed_steps: PlanStep[]
  report_rubric: ReportRubric
  risk_summary: string[]
  approval_required: boolean
  repo_candidates: RepoCandidate[]
  created_at: string
}

export interface ClarificationAnswer {
  question_id: string
  answer: string
}

export interface ClarificationsRequest {
  answers: ClarificationAnswer[]
}

export interface ApprovePlanRequest {
  rubric_weights?: Record<string, number> | null
  disabled_step_ids?: string[]
}

export interface ApprovePlanResponse {
  plan: Plan
  status: string
}

export interface ReplanRequest {
  reason?: string | null
}

export function rubricWeightsSum(weights: Record<string, number>): number {
  return Object.values(weights).reduce((a, b) => a + b, 0)
}

export function isRubricValid(weights: Record<string, number>): boolean {
  return Math.abs(rubricWeightsSum(weights) - 1.0) <= 0.01
}

File: apps/api/mo_api/models/plan.py
"""PlanMode 数据模型（PRD F-002）。"""

from __future__ import annotations

from datetime import datetime, timezone

from pydantic import BaseModel, Field, field_validator, model_validator

from .enums import PlanStepStatus, PlanStepTool, RiskLevel
from .repo_discovery import RepoCandidate

DEFAULT_RUBRIC_WEIGHTS: dict[str, float] = {
    "reproducibility": 0.30,
    "documentation": 0.20,
    "research_value": 0.20,
    "engineering_fit": 0.20,
    "extensibility": 0.10,
}


class ClarifyingQuestion(BaseModel):
    id: str
    question: str
    options: list[str] = Field(default_factory=list)
    answer: str | None = None
    required: bool = False


class PlanStep(BaseModel):
    id: str
    node_id: str  # 对应 ExecuteMode 图中 WorkflowNode 的节点 ID（F-002）
    title: str
    description: str
    tool: PlanStepTool
    risk_level: RiskLevel
    requires_approval: bool
    expected_outputs: list[str] = Field(default_factory=list)
    depends_on: list[str] = Field(default_factory=list)
    user_editable: bool = True
    status: PlanStepStatus = PlanStepStatus.PENDING


class ReportRubric(BaseModel):
    weights: dict[str, float] = Field(default_factory=lambda: dict(DEFAULT_RUBRIC_WEIGHTS))

    @model_validator(mode="after")
    def _weights_sum(self) -> ReportRubric:
        total = sum(self.weights.values())
        if abs(total - 1.0) > 0.01:
            raise ValueError(f"rubric weights must sum to ~1.0, got {total}")
        return self


class Plan(BaseModel):
    id: str
    task_id: str
    task_summary: str
    confirmed_context: list[str] = Field(default_factory=list)
    unknowns: list[str] = Field(default_factory=list)
    clarifying_questions: list[ClarifyingQuestion] = Field(default_factory=list)
    proposed_steps: list[PlanStep] = Field(default_factory=list)
    report_rubric: ReportRubric = Field(default_factory=ReportRubric)
    risk_summary: list[str] = Field(default_factory=list)
    approval_required: bool = True
    # F-015：候选仓库清单（PlanMode 中供用户确认后写回 task.repo_urls）
    repo_candidates: list[RepoCandidate] = Field(default_factory=list)
    created_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))


class ClarificationAnswer(BaseModel):
    question_id: str
    answer: str = Field(min_length=1)


class ClarificationsRequest(BaseModel):
    answers: list[ClarificationAnswer] = Field(min_length=1)


class ApprovePlanRequest(BaseModel):
    rubric_weights: dict[str, float] | None = None
    disabled_step_ids: list[str] = Field(default_factory=list)

    @field_validator("rubric_weights")
    @classmethod
    def _validate_weights(cls, v: dict[str, float] | None) -> dict[str, float] | None:
        if v is None:
            return v
        total = sum(v.values())
        if abs(total - 1.0) > 0.01:
            raise ValueError(f"rubric weights must sum to ~1.0, got {total}")
        return v


class ReplanRequest(BaseModel):
    reason: str | None = None


class PlanResponse(Plan):
    """API 响应：与 Plan 相同。"""


class ApprovePlanResponse(BaseModel):
    plan: Plan
    status: str
